### PR TITLE
Filter by announcement type

### PR DESCRIPTION
**Pull request by hugoabonizio**
Following the idea from #5, this adds a filter by announcement type when clicking on the tag name.

### CODE REVIEW

File: src/models/announcement.cr [hunk 1]
```diff
@@ -36,20 +36,31 @@ class Announcement < Granite::ORM
   validate :description, "is too long",
     ->(this : Announcement) { this.description.to_s.size <= 4000 }
 
-  def self.search(query, per_page = nil, page = 1)
-    self.all %q{
-      WHERE title ILIKE $1 OR description ILIKE $1
+  def self.search(query, per_page = nil, page = 1, type = nil)
+    q = %Q{
+      WHERE (title ILIKE $1 OR description ILIKE $1) #{(type ? "AND type = $4" : "")}
```
Comment (veelenga) [inline src/models/announcement.cr]: Probably this would be a bit cleaner: `#{"AND type = $4" if type}`

File: src/models/announcement.cr [hunk 2]
```diff
@@ -36,20 +36,31 @@ class Announcement < Granite::ORM
...
       ORDER BY created_at DESC
       LIMIT $2 OFFSET $3
-    }, ["%#{query}%", per_page, (page - 1) * per_page]
+    }
+    parameters = if type
```
Comment (veelenga) [inline src/models/announcement.cr]:
```crystal
parameters = ["%#{query}%", per_page, (page - 1) * per_page]
parameters << type if type
```

File: src/models/announcement.cr [hunk 3]
```diff
@@ -36,20 +36,31 @@ class Announcement < Granite::ORM
...
+                   ["%#{query}%", per_page, (page - 1) * per_page, type]
+                 else
+                   ["%#{query}%", per_page, (page - 1) * per_page]
+                 end
+    self.all q, parameters
   end
 
-  def self.count(query)
+  def self.count(query, type = nil)
+    parameters = if type
```
Comment (veelenga) [inline src/models/announcement.cr]:
```crystal
parameters = ["%#{query}%"] of String | Int32
parameters << type if type
```

File: src/models/announcement.cr [hunk 4]
```diff
@@ -36,20 +36,31 @@ class Announcement < Granite::ORM
...
+                 else
+                   "%#{query}%"
+                 end
     @@adapter.open do |db|
-      db.scalar(%q{
+      db.scalar(%Q{
         SELECT COUNT(*) FROM announcements
-        WHERE title ILIKE $1 OR description ILIKE $1
-      }, "%#{query}%").as(Int64)
+        WHERE (title ILIKE $1 OR description ILIKE $1) #{(type ? "AND type = $2" : "")}
```
Comment (veelenga) [inline src/models/announcement.cr]: `#{"AND type = $2" if type}`

File: src/models/announcement.cr [hunk 5]
```diff
@@ -83,6 +94,16 @@ class Announcement < Granite::ORM
...
   end
 
+  def typename_underscore
+    typename.tr(" ", "").underscore
+  end
+
+  def self.reverse_typename_underscore(typename)
+    index = nil
+    TYPES.each { |i, type| index = i if type.tr(" ", "").underscore == typename }
+    index
```
Comment (veelenga) [inline src/models/announcement.cr]: `TYPES.find { |_, type| type.tr(" ", "").underscore == typename }.try &.first`

**Review by veelenga — CHANGES_REQUESTED** [review]
Thanks a lot for this pull request. I have few comments, but that is mostly matter of style than a required change. 

By the way, can we change `Announcement::TYPES` to

```crystal
  TYPES = {
    0 => "blog_post",
    1 => "project_update",
    2 => "conference",
    3 => "meetup",
    4 => "podcast",
    5 => "screencast",
    6 => "video",
    7 => "other",
  }
```
and just adjust `Announcement#typename`, so we will not need to add `Announcement#typename_underscore` and `Announcement.reverse_typename_underscore` methods?

I created new [Gitter channel](https://gitter.im/veelenga/crystal-ann). Let me know if you need to discuss something...

**Comment by hugoabonizio** [comment]
You are right! I made the requested changes. What do you think now?

**Comment by veelenga** [comment]
Looks nice. I will test it this evening and deploy if everything is fine. Thanks again 😻 😼

**Review by veelenga — APPROVED** [review]

**Comment by veelenga** [comment]
Did two small fixed and deployed 👍  

If you plan to continue to work on this project, please let me know the task you are going to do, so we will not do the same job concurrently. See you.

**Comment by hugoabonizio** [comment]
Thanks! Next time I'll let you know throught the gitter channel :smile:

**Pull request merged**